File: finance/backend/src/scripts/import_excel_data.py
#!/usr/bin/env python3
"""
Script para importar dados da planilha Excel do Instituto Areluna para o Supabase
"""

import pandas as pd
import os
import sys
from supabase import create_client, Client
from datetime import datetime
import json
from dotenv import load_dotenv

# Carregar variáveis de ambiente
load_dotenv()

# Configuração do Supabase
SUPABASE_URL = os.getenv('SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_ANON_KEY')

if not SUPABASE_URL or not SUPABASE_KEY:
    print("Erro: Variáveis de ambiente SUPABASE_URL e SUPABASE_ANON_KEY são obrigatórias")
    sys.exit(1)

# Inicializar cliente Supabase
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

def read_excel_file(file_path):
    """
    Lê a planilha Excel e retorna os dados
    """
    try:
        # Ler todas as abas da planilha
        excel_file = pd.ExcelFile(file_path)
        print(f"Abas encontradas: {excel_file.sheet_names}")
        
        sheets_data = {}
        for sheet_name in excel_file.sheet_names:
            df = pd.read_excel(file_path, sheet_name=sheet_name)
            sheets_data[sheet_name] = df
            print(f"\nAba '{sheet_name}': {len(df)} linhas, {len(df.columns)} colunas")
            print(f"Colunas: {list(df.columns)}")
            
        return sheets_data
    except Exception as e:
        print(f"Erro ao ler planilha: {e}")
        return None

def clean_data(df):
    """
    Limpa e padroniza os dados
    """
    # Remover linhas completamente vazias
    df = df.dropna(how='all')
    
    # Converter datas para formato ISO
    for col in df.columns:
        if df[col].dtype == 'datetime64[ns]':
            df[col] = df[col].dt.strftime('%Y-%m-%d')
    
    # Substituir NaN por None
    df = df.where(pd.notnull(df), None)
    
    return df

def process_clients_data(df):
    """
    Processa dados de clientes
    """
    clients = []
    
    for _, row in df.iterrows():
        # Mapear colunas da planilha para campos do banco
        client = {
            'first_name': str(row.get('Nome', '')).strip() if pd.notna(row.get('Nome')) else 'Cliente',
            'last_name': str(row.get('Sobrenome', '')).strip() if pd.notna(row.get('Sobrenome')) else '',
            'email': str(row.get('Email', '')).strip() if pd.notna(row.get('Email')) else None,
            'phone': str(row.get('Telefone', '')).strip() if pd.notna(row.get('Telefone')) else None,
            'mobile': str(row.get('Celular', '')).strip() if pd.notna(row.get('Celular')) else None,
            'tax_id': str(row.get('CPF', '')).strip() if pd.notna(row.get('CPF')) else None,
            'birth_date': row.get('Data_Nascimento') if pd.notna(row.get('Data_Nascimento')) else None,
            'address': str(row.get('Endereco', '')).strip() if pd.notna(row.get('Endereco')) else None,
            'city': str(row.get('Cidade', '')).strip() if pd.notna(row.get('Cidade')) else None,
            'postal_code': str(row.get('CEP', '')).strip() if pd.notna(row.get('CEP')) else None,
            'country': 'Portugal',
            'attention_level': 'normal',
            'notes': str(row.get('Observacoes', '')).strip() if pd.notna(row.get('Observacoes')) else None,
            'is_active': True,
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat()
        }
        
        # Validar dados obrigatórios
        if client['first_name']:
            clients.append(client)
    
    return clients

def process_contracts_data(df, clients_map):
    """
    Processa dados de contratos
    """
    contracts = []
    
    for _, row in df.iterrows():
        # Encontrar cliente pelo nome ou email
        client_id = None
        client_name = str(row.get('Cliente', '')).strip() if pd.notna(row.get('Cliente')) else ''
        
        if client_name and client_name in clients_map:
            client_id = clients_map[client_name]
        
        if not client_id:
            continue
            
        contract = {
            'client_id': client_id,
            'contract_number': str(row.get('Numero_Contrato', '')).strip() if pd.notna(row.get('Numero_Contrato')) else f"CONT-{datetime.now().strftime('%Y%m%d')}-{len(contracts)+1:03d}",
            'description': str(row.get('Descricao', '')).strip() if pd.notna(row.get('Descricao')) else 'Contrato de Serviços',
            'total_amount': float(row.get('Valor_Total', 0)) if pd.notna(row.get('Valor_Total')) else 0.0,
            'start_date': row.get('Data_Inicio') if pd.notna(row.get('Data_Inicio')) else datetime.now().strftime('%Y-%m-%d'),
            'end_date': row.get('Data_Fim') if pd.notna(row.get('Data_Fim')) else None,
            'status': str(row.get('Status', 'active')).lower() if pd.notna(row.get('Status')) else 'active',
            'payment_frequency': str(row.get('Frequencia_Pagamento', 'monthly')).lower() if pd.notna(row.get('Frequencia_Pagamento')) else 'monthly',
            'notes': str(row.get('Observacoes', '')).strip() if pd.notna(row.get('Observacoes')) else None,
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat()
        }
        
        contracts.append(contract)
    
    return contracts

def process_payments_data(df, contracts_map):
    """
    Processa dados de pagamentos
    """
    payments = []
    
    for _, row in df.iterrows():
        # Encontrar contrato pelo número
        contract_id = None
        contract_number = str(row.get('Numero_Contrato', '')).strip() if pd.notna(row.get('Numero_Contrato')) else ''
        
        if contract_number and contract_number in contracts_map:
            contract_id = contracts_map[contract_number]
        
        if not contract_id:
            continue
            
        payment = {
            'contract_id': contract_id,
            'amount': float(row.get('Valor', 0)) if pd.notna(row.get('Valor')) else 0.0,
            'due_date': row.get('Data_Vencimento') if pd.notna(row.get('Data_Vencimento')) else datetime.now().strftime('%Y-%m-%d'),
            'paid_date': row.get('Data_Pagamento') if pd.notna(row.get('Data_Pagamento')) else None,
            'status': str(row.get('Status_Pagamento', 'pending')).lower() if pd.notna(row.get('Status_Pagamento')) else 'pending',
            'payment_method': str(row.get('Metodo_Pagamento', 'bank_transfer')).lower() if pd.notna(row.get('Metodo_Pagamento')) else 'bank_transfer',
            'notes': str(row.get('Observacoes', '')).strip() if pd.notna(row.get('Observacoes')) else None,
            'created_at': datetime.now().isoformat(),
            'updated_at': datetime.now().isoformat()
        }
        
        payments.append(payment)
    
    return payments

def insert_data_to_supabase(table_name, data):
    """
    Insere dados no Supabase
    """
    try:
        if not data:
            print(f"Nenhum dado para inserir na tabela {table_name}")
            return []
            
        print(f"Inserindo {len(data)} registros na tabela {table_name}...")
        
        # Inserir em lotes de 100
        batch_size = 100
        inserted_records = []
        
        for i in range(0, len(data), batch_size):
            batch = data[i:i + batch_size]
            result = supabase.table(table_name).insert(batch).execute()
            
            if result.data:
                inserted_records.extend(result.data)
                print(f"Lote {i//batch_size + 1}: {len(result.data)} registros inseridos")
            else:
                print(f"Erro ao inserir lote {i//batch_size + 1}: {result}")
        
        print(f"Total de {len(inserted_records)} registros inseridos na tabela {table_name}")
        return inserted_records
        
    except Exception as e:
        print(f"Erro ao inserir dados na tabela {table_name}: {e}")
        return []

def main():
    """
    Função principal
    """
    excel_file_path = "/Users/insitutoareluna/Documents/finance/MACRO_PAGAMENTOS INSTITUTO ARELUNA - Editável (1).xlsm"
    
    if not os.path.exists(excel_file_path):
        print(f"Arquivo não encontrado: {excel_file_path}")
        return
    
    print("Iniciando importação de dados da planilha Excel...")
    
    # Ler planilha
    sheets_data = read_excel_file(excel_file_path)
    if not sheets_data:
        return
    
    # Processar cada aba
    clients_map = {}
    contracts_map = {}
    
    for sheet_name, df in sheets_data.items():
        print(f"\n=== Processando aba: {sheet_name} ===")
        
        # Limpar dados
        df_clean = clean_data(df)
        
        # Mostrar amostra dos dados
        print("\nPrimeiras 3 linhas:")
        print(df_clean.head(3).to_string())
        
        # Processar baseado no nome da aba ou conteúdo
        if 'client' in sheet_name.lower() or 'cliente' in sheet_name.lower():
            clients_data = process_clients_data(df_clean)
            inserted_clients = insert_data_to_supabase('clients', clients_data)
            
            # Criar mapeamento de clientes
            for client in inserted_clients:
                full_name = f"{client['first_name']} {client['last_name']}".strip()
                clients_map[full_name] = client['id']
                
        elif 'contract' in sheet_name.lower() or 'contrato' in sheet_name.lower():
            contracts_data = process_contracts_data(df_clean, clients_map)
            inserted_contracts = insert_data_to_supabase('contracts', contracts_data)
            
            # Criar mapeamento de contratos
            for contract in inserted_contracts:
                contracts_map[contract['contract_number']] = contract['id']
                
        elif 'payment' in sheet_name.lower() or 'pagamento' in sheet_name.lower():
            payments_data = process_payments_data(df_clean, contracts_map)
            insert_data_to_supabase('payments', payments_data)
        
        else:
            print(f"Aba '{sheet_name}' não reconhecida. Tentando processar como dados gerais...")
            # Tentar identificar pelo conteúdo das colunas
            columns = [col.lower() for col in df_clean.columns]
            
            if any(col in ['nome', 'email', 'telefone', 'cpf'] for col in columns):
                print("Detectados dados de clientes")
                clients_data = process_clients_data(df_clean)
                inserted_clients = insert_data_to_supabase('clients', clients_data)
                
                for client in inserted_clients:
                    full_name = f"{client['first_name']} {client['last_name']}".strip()
                    clients_map[full_name] = client['id']
    
    print("\n=== Importação concluída ===")
    print(f"Clientes mapeados: {len(clients_map)}")
    print(f"Contratos mapeados: {len(contracts_map)}")

if __name__ == "__main__":
    main()

File: backend/src/scripts/import_excel_data.js
#!/usr/bin/env node

/**
 * Script para importar dados da planilha Excel do Instituto Areluna para o Supabase
 */

const XLSX = require('xlsx');
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
require('dotenv').config();

// Configuração do Supabase
const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseKey) {
    console.error('❌ Erro: Variáveis de ambiente SUPABASE_URL e SUPABASE_ANON_KEY são obrigatórias');
    console.log('Por favor, configure o arquivo .env com as credenciais do Supabase');
    process.exit(1);
}

// Inicializar cliente Supabase
const supabase = createClient(supabaseUrl, supabaseKey);

/**
 * Processa dados de clientes da aba "Banco de Dados"
 */
function processClientsFromBancoDados(data) {
    console.log('📋 Processando dados de clientes da aba "Banco de Dados"...');
    
    const clients = [];
    
    data.forEach((row, index) => {
        // Pular linhas vazias
        if (!row.Nome || row.Nome.toString().trim() === '') {
            return;
        }
        
        const fullName = row.Nome.toString().trim();
        const nameParts = fullName.split(' ');
        const firstName = nameParts[0] || 'Cliente';
        const lastName = nameParts.slice(1).join(' ') || '';
        
        const client = {
            first_name: firstName,
            last_name: lastName,
            email: row['E-mail'] ? row['E-mail'].toString().trim() : null,
            phone: null, // Não disponível nesta aba
            mobile: null, // Não disponível nesta aba
            tax_id: null, // Não disponível nesta aba
            birth_date: null, // Não disponível nesta aba
            address: row.Morada ? row.Morada.toString().trim() : null,
            city: row.Cidade ? row.Cidade.toString().trim() : null,
            postal_code: row['Código postal'] ? row['Código postal'].toString().trim() : null,
            country: 'Portugal',
            attention_level: 'normal',
            notes: row.IBAN ? `IBAN: ${row.IBAN}` : null,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
        };
        
        // Validar email se fornecido
        if (client.email && !client.email.includes('@')) {
            client.email = null;
        }
        
        clients.push(client);
    });
    
    console.log(`✅ ${clients.length} clientes processados`);
    return clients;
}

/**
 * Processa dados de contratos da aba "Contratos Ativos"
 */
function processContractsFromAtivos(data, clientsMap) {
    console.log('📋 Processando dados de contratos da aba "Contratos Ativos"...');
    
    const contracts = [];
    
    data.forEach((row, index) => {
        // Verificar se há dados suficientes
        if (!row.Nome || row.Nome.toString().trim() === '') {
            return;
        }
        
        const clientName = row.Nome.toString().trim();
        let clientId = null;
        
        // Procurar cliente no mapeamento
        if (clientsMap[clientName]) {
            clientId = clientsMap[clientName];
        } else {
            // Tentar busca parcial
            const foundClient = Object.keys(clientsMap).find(name => 
                name.toLowerCase().includes(clientName.toLowerCase()) ||
                clientName.toLowerCase().includes(name.toLowerCase())
            );
            if (foundClient) {
                clientId = clientsMap[foundClient];
            }
        }
        
        if (!clientId) {
            console.warn(`⚠️  Cliente não encontrado para contrato: ${clientName}`);
            return;
        }
        
        const contract = {
            client_id: clientId,
            contract_number: row['Nº Contrato'] ? row['Nº Contrato'].toString().trim() : `CONT-${new Date().toISOString().slice(0, 10).replace(/-/g, '')}-${String(index + 1).padStart(3, '0')}`,
            description: row.Curso ? `Curso: ${row.Curso}` : 'Contrato de Serviços',
            total_amount: parseFloat(row['Valor Total'] || row['Valor'] || 0),
            start_date: row['Data Início'] || new Date().toISOString().split('T')[0],
            end_date: row['Data Fim'] || null,
            status: 'active',
            payment_frequency: 'monthly',
            notes: null,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
        };
        
        // Processar datas se estiverem no formato Excel
        if (typeof contract.start_date === 'number') {
            const excelDate = new Date((contract.start_date - 25569) * 86400 * 1000);
            if (!isNaN(excelDate.getTime())) {
                contract.start_date = excelDate.toISOString().split('T')[0];
            }
        }
        
        contracts.push(contract);
    });
    
    console.log(`✅ ${contracts.length} contratos processados`);
    return contracts;
}

/**
 * Processa dados de pagamentos da aba "PAGAMENTOS PARCIAIS"
 */
function processPaymentsFromParciais(data, contractsMap) {
    console.log('📋 Processando dados de pagamentos da aba "PAGAMENTOS PARCIAIS"...');
    
    const payments = [];
    
    data.forEach((row, index) => {
        // Verificar se há dados suficientes
        if (!row['Nº Contrato'] && !row.Nome) {
            return;
        }
        
        let contractId = null;
        const contractNumber = row['Nº Contrato'] ? row['Nº Contrato'].toString().trim() : null;
        
        if (contractNumber && contractsMap[contractNumber]) {
            contractId = contractsMap[contractNumber];
        }
        
        if (!contractId) {
            console.warn(`⚠️  Contrato não encontrado para pagamento: ${contractNumber || row.Nome}`);
            return;
        }
        
        const payment = {
            contract_id: contractId,
            amount: parseFloat(row.Valor || row['Valor Pago'] || 0),
            due_date: row['Data Vencimento'] || new Date().toISOString().split('T')[0],
            paid_date: row['Data Pagamento'] || null,
            status: row['Data Pagamento'] ? 'paid' : 'pending',
            payment_method: 'bank_transfer',
            notes: row.Observações || null,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
        };
        
        // Processar datas se estiverem no formato Excel
        ['due_date', 'paid_date'].forEach(dateField => {
            if (typeof payment[dateField] === 'number') {
                const excelDate = new Date((payment[dateField] - 25569) * 86400 * 1000);
                if (!isNaN(excelDate.getTime())) {
                    payment[dateField] = excelDate.toISOString().split('T')[0];
                }
            }
        });
        
        payments.push(payment);
    });
    
    console.log(`✅ ${payments.length} pagamentos processados`);
    return payments;
}

/**
 * Insere dados no Supabase
 */
async function insertDataToSupabase(tableName, data) {
    try {
        if (!data || data.length === 0) {
            console.log(`ℹ️  Nenhum dado para inserir na tabela ${tableName}`);
            return [];
        }
        
        console.log(`🚀 Inserindo ${data.length} registros na tabela ${tableName}...`);
        
        // Inserir em lotes de 50 para evitar timeouts
        const batchSize = 50;
        const insertedRecords = [];
        
        for (let i = 0; i < data.length; i += batchSize) {
            const batch = data.slice(i, i + batchSize);
            
            console.log(`   📦 Processando lote ${Math.floor(i / batchSize) + 1}/${Math.ceil(data.length / batchSize)}...`);
            
            const { data: result, error } = await supabase
                .from(tableName)
                .insert(batch)
                .select();
            
            if (error) {
                console.error(`❌ Erro ao inserir lote ${Math.floor(i / batchSize) + 1}:`, error.message);
                
                // Tentar inserir registros individualmente para identificar problemas
                for (const record of batch) {
                    try {
                        const { data: singleResult, error: singleError } = await supabase
                            .from(tableName)
                            .insert([record])
                            .select();
                        
                        if (singleError) {
                            console.error(`❌ Erro no registro:`, singleError.message, record);
                        } else if (singleResult) {
                            insertedRecords.push(...singleResult);
                        }
                    } catch (singleErr) {
                        console.error(`❌ Erro crítico no registro:`, singleErr.message);
                    }
                }
                continue;
            }
            
            if (result) {
                insertedRecords.push(...result);
                console.log(`   ✅ ${result.length} registros inseridos`);
            }
        }
        
        console.log(`🎉 Total de ${insertedRecords.length} registros inseridos na tabela ${tableName}`);
        return insertedRecords;
        
    } catch (error) {
        console.error(`❌ Erro crítico ao inserir dados na tabela ${tableName}:`, error.message);
        return [];
    }
}

/**
 * Função principal
 */
async function main() {
    const excelFilePath = '/Users/pedro/Documents/finance/MACRO_PAGAMENTOS INSTITUTO ARELUNA - Editável (1).xlsm';
    
    console.log('🚀 === Iniciando Importação de Dados da Planilha Excel ===');
    console.log(`📁 Arquivo: ${excelFilePath}`);
    
    if (!fs.existsSync(excelFilePath)) {
        console.error(`❌ Arquivo não encontrado: ${excelFilePath}`);
        return;
    }
    
    try {
        // Ler planilha
        console.log('📖 Lendo arquivo Excel...');
        const workbook = XLSX.readFile(excelFilePath);
        console.log(`✅ Arquivo lido. Abas encontradas: ${workbook.SheetNames.length}`);
        console.log(`📊 Nomes das abas: ${workbook.SheetNames.join(', ')}`);
        
        // Mapear dados das abas principais
        const clientsMap = {};
        const contractsMap = {};
        
        // 1. Processar clientes da aba "Banco de Dados"
        if (workbook.SheetNames.includes('Banco de Dados')) {
            console.log('\n📊 === Processando Clientes ===');
            console.log('📋 Aba "Banco de Dados" encontrada');
            const bancoDadosSheet = workbook.Sheets['Banco de Dados'];
            const bancoDadosData = XLSX.utils.sheet_to_json(bancoDadosSheet, { defval: null });
            console.log(`📊 ${bancoDadosData.length} linhas encontradas na aba`);
            
            const clientsData = processClientsFromBancoDados(bancoDadosData);
            console.log('💾 Inserindo clientes no Supabase...');
            const insertedClients = await insertDataToSupabase('clients', clientsData);
            
            // Criar mapeamento de clientes
            insertedClients.forEach(client => {
                const fullName = `${client.first_name} ${client.last_name}`.trim();
                clientsMap[fullName] = client.id;
            });
            
            console.log(`🗂️  Clientes mapeados: ${Object.keys(clientsMap).length}`);
        } else {
            console.log('⚠️  Aba "Banco de Dados" não encontrada');
        }
        
        // 2. Processar contratos da aba "Contratos Ativos"
        if (workbook.SheetNames.includes('Contratos Ativos')) {
            console.log('\n📊 === Processando Contratos ===');
            console.log('📋 Aba "Contratos Ativos" encontrada');
            const contratosSheet = workbook.Sheets['Contratos Ativos'];
            const contratosData = XLSX.utils.sheet_to_json(contratosSheet, { defval: null });
            console.log(`📊 ${contratosData.length} linhas encontradas na aba`);
            
            const contractsData = processContractsFromAtivos(contratosData, clientsMap);
            console.log('💾 Inserindo contratos no Supabase...');
            const insertedContracts = await insertDataToSupabase('contracts', contractsData);
            
            // Criar mapeamento de contratos
            insertedContracts.forEach(contract => {
                contractsMap[contract.contract_number] = contract.id;
            });
            
            console.log(`🗂️  Contratos mapeados: ${Object.keys(contractsMap).length}`);
        } else {
            console.log('⚠️  Aba "Contratos Ativos" não encontrada');
        }
        
        // 3. Processar pagamentos da aba "PAGAMENTOS PARCIAIS"
        if (workbook.SheetNames.includes('PAGAMENTOS PARCIAIS')) {
            console.log('\n📊 === Processando Pagamentos ===');
            console.log('📋 Aba "PAGAMENTOS PARCIAIS" encontrada');
            const pagamentosSheet = workbook.Sheets['PAGAMENTOS PARCIAIS'];
            const pagamentosData = XLSX.utils.sheet_to_json(pagamentosSheet, { defval: null });
            console.log(`📊 ${pagamentosData.length} linhas encontradas na aba`);
            
            const paymentsData = processPaymentsFromParciais(pagamentosData, contractsMap);
            console.log('💾 Inserindo pagamentos no Supabase...');
            await insertDataToSupabase('payments', paymentsData);
        } else {
            console.log('⚠️  Aba "PAGAMENTOS PARCIAIS" não encontrada');
        }
        
        console.log('\n🎉 === Importação Concluída com Sucesso! ===');
        console.log(`👥 Clientes importados: ${Object.keys(clientsMap).length}`);
         console.log(`📄 Contratos importados: ${Object.keys(contractsMap).length}`);
         console.log('📊 Resumo do processamento completo');
         
         // Salvar mapeamentos para referência
        const mappings = {
            clients: clientsMap,
            contracts: contractsMap,
            importDate: new Date().toISOString()
        };
        
        fs.writeFileSync('/Users/pedro/Documents/finance/backend/import_mappings.json', JSON.stringify(mappings, null, 2));
        console.log('💾 Mapeamentos salvos em import_mappings.json');
        
    } catch (error) {
        console.error('❌ Erro durante a importação:', error.message);
        console.error('Stack trace:', error.stack);
    }
}

// Executar se chamado diretamente
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { main };